Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp

```diff
@@ -0,0 +1,144 @@
+
+    // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
+    if (path.startsWith('/'))
+        path = path.substring(1);
+    
+    if (auto cachedObject = m_resources.getOptional(path)) {
+        if (auto cachedString = String::fromUTF8(*cachedObject); !cachedString.isEmpty())
+            return cachedString.utf8().span();
+        
+        ASSERT(!cachedObject.isEmpty());
+        
+        return *cachedObject;
+    }
+
+    bool isServiceWorker = backgroundContentIsServiceWorker();
+    if (!isServiceWorker && (path == generatedBackgroundPageFilename))
+        return generatedBackgroundContent().utf8().span();
+    
+    if (isServiceWorker && (path == generatedBackgroundServiceWorkerFilename))
+        return generatedBackgroundContent().utf8().span();
+    
+    URL *resourceURL = resourceFileURLForPath(path);
+    if (!resourceURL) {
+        if (suppressErrors == SuppressNotFoundErrors::No)
+            recordError(createError(Error::ResourceNotFound, WEB_UI_FORMAT_STRING("Unable to find \"%s\" in the extension’s resources. It is an invalid path.", "WKWebExtensionErrorResourceNotFound description with invalid file path", path.utf8().data())));
+        return std::span<const uint8_t>();
+    }
+
+    auto data = FileSystem::readEntireFile(resourceURL->fileSystemPath());
+    if (!data.has_value()) {
+        if (suppressErrors == SuppressNotFoundErrors::No)
+            recordError(createError(Error::ResourceNotFound, WEB_UI_FORMAT_CFSTRING("Unable to find \"%@\" in the extension’s resources.", "WKWebExtensionErrorResourceNotFound description with file name", path.utf8().data())));
+        return std::span<const uint8_t>();
+    }
+
+    if (cacheResult == CacheResult::Yes)
+        m_resources.set(String(path), (*data).span());
+
+    return *data;
+}
+
+void WebExtension::recordError(Ref<API::Error> error, SuppressNotification suppressNotifications)
+{
+    if (!m_errors.isEmpty())
+        m_errors = Vector<Ref<API::Error>>();
+    
+    RELEASE_LOG_ERROR(Extensions, "Error recorded: %s", error.get().localizedDescription().utf8().data());
+
+    m_errors.append(error);
+
+    if (suppressNotifications == SuppressNotification::Yes)
+        return;
+    
+    // FIXME: dispatch errors on Linux
+}
+
+RefPtr<WebCore::Icon> WebExtension::imageForPath(String path)
+{
+    ASSERT(path);
+
+    auto imageData = resourceDataForPath(path, CacheResult::Yes);
+    if (!imageData.empty())
+        return nullptr;
+
+    GBytes *imageBytes = g_bytes_new (imageData.data(), imageData.size());
+    GIcon *image = g_bytes_icon_new (imageBytes);
+
+    return WebCore::Icon::create(image);
+}
+
+RefPtr<WebCore::Icon> WebExtension::bestImageInIconsDictionary(RefPtr<JSON::Value> value, size_t idealPointSize)
+{
+    if (!value->asObject()->size())
+        return nullptr;
+
+    int standardPixelSize = idealPointSize;
+
+    auto standardIconPath = pathForBestImageInIconsDictionary(value, standardPixelSize);
```

Comment:
I'd probably have `pathForBestImageInIconsDictionary()` also return a `std::optional<String>` and never an empty string.